File: core/services/proyecciones_service.py
import math
from psycopg.rows import dict_row
from ..db import get_connection

SMA_WINDOW = 3


def mean(arr: list) -> float:
    return sum(arr) / len(arr)


def std_dev(arr: list) -> float:
    m = mean(arr)
    variance = sum((v - m) ** 2 for v in arr) / len(arr)
    return math.sqrt(variance)


def _add_month(fecha_yyyy_mm: str) -> str:
    year, month = map(int, fecha_yyyy_mm.split("-"))
    month += 1
    if month > 12:
        month = 1
        year += 1
    return f"{year}-{month:02d}"


def get_provincias() -> list:
    with get_connection() as conn:
        with conn.cursor() as cur:
            cur.execute("SELECT nombre FROM betix.provincias ORDER BY nombre")
            return [r[0] for r in cur.fetchall()]


def get_juegos() -> list:
    with get_connection() as conn:
        with conn.cursor() as cur:
            cur.execute("SELECT nombre FROM betix.juegos ORDER BY nombre")
            return [r[0] for r in cur.fetchall()]


def calcular_proyecciones(provincia: str, juego: str, k: int, n: int = SMA_WINDOW) -> dict:
    with get_connection() as conn:
        with conn.cursor(row_factory=dict_row) as cur:
            cur.execute("""
                SELECT
                    TO_CHAR(t.fecha, 'YYYY-MM') AS fecha,
                    t.cantidad,
                    t.ingresos,
                    t.costo,
                    (t.ingresos - t.costo)      AS beneficio
                FROM betix.tickets_mensuales t
                JOIN betix.provincias p ON p.id = t.provincia_id
                JOIN betix.juegos     j ON j.id = t.juego_id
                WHERE p.nombre = %s
                  AND j.nombre = %s
                ORDER BY t.fecha
            """, (provincia, juego))
            historico = [
                {
                    "fecha":     r["fecha"],
                    "cantidad":  r["cantidad"],
                    "ingresos":  int(r["ingresos"]),
                    "costo":     int(r["costo"]),
                    "beneficio": int(r["beneficio"]),
                }
                for r in cur.fetchall()
            ]

    if len(historico) < n:
        raise ValueError(
            f"Datos insuficientes para proyectar (se necesitan {n} meses, hay {len(historico)})"
        )

    metricas = ["cantidad", "ingresos", "costo", "beneficio"]

    series   = {met: [h[met] for h in historico] for met in metricas}
    base_sds = {met: std_dev([h[met] for h in historico][-n:]) for met in metricas}

    last_fecha = historico[-1]["fecha"]
    proyectado = []

    for i in range(k):
        fecha = _add_month(last_fecha)
        entry = {"fecha": fecha}

        for met in metricas:
            window = series[met][-n:]
            valor  = round(mean(window))
            error  = round(base_sds[met] * (1 + i * 0.15))

            entry[met]            = valor
            entry[f"error_{met}"] = error
            series[met]           = series[met] + [valor]

        proyectado.append(entry)
        last_fecha = fecha

    return {"historico": historico, "proyectado": proyectado}

File: src/services/proyeccionesService.js
'use strict';

const { ticketsPorMes } = require('../data/ticketsPorMes');

const SMA_WINDOW = 3; // ventana por defecto para la media móvil simple

// ── Helpers estadísticos ──────────────────────────────────────────────────────

function mean(arr) {
  return arr.reduce((a, b) => a + b, 0) / arr.length;
}

function stdDev(arr) {
  const m = mean(arr);
  const variance = arr.reduce((sum, v) => sum + (v - m) ** 2, 0) / arr.length;
  return Math.sqrt(variance);
}

function addMonth(fechaYYYYMM) {
  const [y, m] = fechaYYYYMM.split('-').map(Number);
  const d = new Date(y, m - 1 + 1, 1);
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`;
}

// ── Proyecciones ──────────────────────────────────────────────────────────────

/**
 * Calcula proyecciones por SMA rolling para una provincia/juego.
 *
 * @param {object} params
 * @param {string} params.provincia
 * @param {string} params.juego
 * @param {number} params.k   meses a proyectar (1–4)
 * @param {number} [params.n] ventana SMA (default: 3)
 * @returns {{ historico: object[], proyectado: object[] }}
 */
function calcularProyecciones({ provincia, juego, k, n = SMA_WINDOW }) {
  const historico = ticketsPorMes
    .filter(t => t.provincia === provincia && t.juego === juego)
    .sort((a, b) => a.fecha.localeCompare(b.fecha))
    .map(t => ({ ...t, beneficio: t.ingresos - t.costo }));

  if (historico.length < n) {
    throw new Error(
      `Datos insuficientes para proyectar (se necesitan ${n} meses, hay ${historico.length})`
    );
  }

  const metricas = ['cantidad', 'ingresos', 'costo', 'beneficio'];

  // Series de trabajo para rolling SMA (se extienden con cada proyectado)
  const series = Object.fromEntries(
    metricas.map(met => [met, historico.map(h => h[met])])
  );

  // SD base calculada sobre el histórico (últimos n meses): garantiza crecimiento monotónico del error
  const baseSDs = Object.fromEntries(
    metricas.map(met => [met, stdDev(historico.map(h => h[met]).slice(-n))])
  );

  let lastFecha = historico[historico.length - 1].fecha;
  const proyectado = [];

  for (let i = 0; i < k; i++) {
    const fecha = addMonth(lastFecha);
    const entry = { fecha };

    for (const met of metricas) {
      const window = series[met].slice(-n);
      const valor  = Math.round(mean(window));
      // Error crece desde la SD histórica base, garantizando monotonía
      const error  = Math.round(baseSDs[met] * (1 + i * 0.15));

      entry[met]             = valor;
      entry[`error_${met}`]  = error;

      series[met] = [...series[met], valor]; // rolling: incluir proyectado para SMA
    }

    proyectado.push(entry);
    lastFecha = fecha;
  }

  return { historico, proyectado };
}

// ── Listas de referencia (ordenadas alfabéticamente) ─────────────────────────

function getProvincias() {
  return [...new Set(ticketsPorMes.map(t => t.provincia))].sort();
}

function getJuegos() {
  return [...new Set(ticketsPorMes.map(t => t.juego))].sort();
}

module.exports = { calcularProyecciones, getProvincias, getJuegos, mean, stdDev };
